Add Non-Uniform Border to Border.

bernaferrari opened the pull request:
Fix https://github.com/flutter/flutter/issues/12583.
Fix https://github.com/flutter/flutter/issues/109954.

Exactly like Figma:

<img width="206" alt="image" src="https://user-images.githubusercontent.com/351125/222874009-ee49ed13-9538-413b-a6da-bba1d4401056.png">

<img width="652" alt="image" src="https://user-images.githubusercontent.com/351125/222873967-e446392a-e2ca-4a35-ad8b-37b951e1ce20.png">

What if it has multiple colors + borderRadius? It will crash as usual.
Side effects: material data tables will paint faster, because drrect > path.

bernaferrari committed:
Add Non-Uniform Border to Border.

bernaferrari committed:
Fix BorderDirectional and dimensions.

bernaferrari committed:
Relax strokeAlign requirements.

bernaferrari committed:
Fix test.

Also add more tests.

bernaferrari committed:
Replace _clampRadius with Radius.clamp.

bernaferrari commented on packages/flutter/lib/src/painting/box_border.dart:
```diff
@@ -567,7 +632,7 @@ class Border extends BoxBorder {
       if (!_strokeAlignIsUniform || top.strokeAlign != BorderSide.strokeAlignInside) {
         throw FlutterError.fromParts(<DiagnosticsNode>[
-          ErrorSummary('A Border can only draw strokeAlign different than BorderSide.strokeAlignInside on uniform borders.'),
+          ErrorSummary('A Border can only draw strokeAlign different than BorderSide.strokeAlignInside on borders with uniform colors.'),
```
`uniform colors/styles` instead? Uniform colors is way more common and `style == BorderStyle.none` is kind of an invisible color.

bernaferrari replied:
Thanks!

bernaferrari commented:
Hey @gspencergoog, did you miss me? 😛

gspencergoog commented on packages/flutter/lib/src/painting/box_border.dart:
```diff
@@ -538,15 +578,25 @@ class Border extends BoxBorder {
       }
     }
 
+    // Allow painting non-uniform borders if the color is uniform.
```
```suggestion
    // Allow painting non-uniform borders if the shape is rectangular, and the color and style are uniform.
```

bernaferrari replied:
I think I'll allow BoxShape.circle. I've been trying to deprecate it for 8 months, but that will give a reason for it to exist.

Let me see if I can implement non-uniform border for circles.

gspencergoog commented on packages/flutter/lib/src/painting/box_border.dart:
```diff
@@ -538,15 +578,25 @@ class Border extends BoxBorder {
+    // Allow painting non-uniform borders if the color is uniform.
+    //
+    // See also:
+    //  * <https://pub.dev/packages/non_uniform_border>, a package that
```
Can you move this "See also" to a real doc comment so people can see it?

bernaferrari replied:
Done, check if you prefer in Border or paint (probably paint, right?).

gspencergoog commented on packages/flutter/test/painting/border_test.dart:
```diff
@@ -299,5 +299,21 @@ void main() {
+      end: BorderSide(width: 15, strokeAlign: BorderSide.strokeAlignOutside),
+      bottom: BorderSide(width: 20),
+    );
+    expect(nonUniformBorderDirectional.dimensions, const EdgeInsetsDirectional.fromSTEB(5, 5, 0, 20));
```
Can you also test to make sure that it asserts when non-uniform borders of various types are tried? You don't have to test for the assert message itself, but at least that it asserts under the right conditions.

bernaferrari replied:
Yessss. Hold on.

bernaferrari commented:
Do you think `BorderStyle.none` could be considered as `width == 0` (right now they are `color == transparent`)? If I make it the same as `width == 0` in non-uniform borders, it is one less exception. Only uniform colors would need to be required.

gspencergoog commented:
>  Only uniform colors would need to be required.

Again, what happens if we add a new style?

bernaferrari commented:
> Again, what happens if we add a new style?

Based on this question, I changed half my implementation to use switch so it is more future-proof 😛.

Ok, so I refactored eeeeeverything (hopefully for the better). I added support for non-uniform circle (it is a shame we can't have ovals) and improved the documentation. I got unsure if you meant the "see also" from Border or from `paint`, so I put on both (feel free to correct me which one you meant).

gspencergoog commented on packages/flutter/lib/src/painting/box_border.dart:
```diff
@@ -212,7 +212,8 @@ abstract class BoxBorder extends ShapeBorder {
   ///
-  ///  * [paintBorder], which is used if the border is not uniform.
+  ///  * [paintBorder], which is used if the border has non-uniform colors or styles and no borderRadius.
+  ///  * [Border.paint], which is the similar method in [Border] with additional comments.
```
What do you mean by "additional comments"? More context? I'd reword this in any case.

bernaferrari replied:
Yes. I'll reword. BorderDirectional is shorter and gives less detail on what each parameter do than the equivalent Border.

bernaferrari replied:
Done. Replacement is:

```dart
[Border.paint], similar to this method, includes additional comments and provides more detailed information on each parameter than described here.
```

gspencergoog commented on packages/flutter/lib/src/painting/box_border.dart:
```diff
@@ -240,6 +241,64 @@ abstract class BoxBorder extends ShapeBorder {
+            .toRRect(rect);
+        break;
+      case BoxShape.circle:
+        assert(borderRadius == null, 'A borderRadius can only be given for rectangular boxes.');
```
```suggestion
        assert(borderRadius == null, 'A borderRadius cannot be given when shape is a BoxShape.circle.');
```

gspencergoog commented on packages/flutter/test/painting/border_test.dart:
```diff
@@ -299,5 +298,108 @@ void main() {
+        borderRadius: BorderRadius.circular(25),
+      ),
+    );
+    expect(tester.takeException(), isNotNull,
```
I'd use `isAssertionError` instead of `isNotNull` to be more specific. Here and below.

bernaferrari replied:
Thanks. I didn't know about this.

gspencergoog reviewed (approved):
![32384589-a60f0e74-c078-11e7-9bc1-e5b5287aea9d](https://user-images.githubusercontent.com/8867023/37737237-66e9c9dc-2d10-11e8-8797-13a405eace49.png)

bernaferrari committed:
Simplify + Fix Dart 3 issues.

bernaferrari commented:
This will have a small, probably wanted, effect. If there is borderRadius + Border.all, the border will not be clipped around anymore (the image below). Will work like RoundedRectangleBorder.

<img width="526" ### alt="image" src="https://user-images.githubusercontent.com/351125/227292608-aca021be-9877-46da-a4e0-a0352320b315.png">

Piinks reviewed (approved):
LGTM
